### PR TITLE
OTelTracingProvider should be request-scoped

### LINKED ISSUE #35427: OTEL: OTelTracingProvider should be request-scoped
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

dist/quarkus

### Describe the bug

The creation of the `OtelTracingProvider.java` is done as application-scoped as the singleton pattern is used to return always the same instance of the provider. However, the provider contains a state, that should not be shared among the application.

### Version

999

### Regression

- [ ] The issue is a regression

### Expected behavior

`OtelTracingProvider` is request-scoped.

### Actual behavior

`OtelTracingProvider` is application-scoped.

### How to Reproduce?

...

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Fixes #35427

Also making the deque of scopes thread-safe.

@ahus1 Could you please check it? Thanks

@brunobat JFYI

### CODE REVIEW

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/tracing/OTelTracingProvider.java [hunk 1]
```diff
@@ -45,7 +45,7 @@ public class OTelTracingProvider implements TracingProvider {
 
     public OTelTracingProvider(OpenTelemetry openTelemetry) {
         this.openTelemetry = openTelemetry;
-        this.scopes = new ArrayDeque<>();
+        this.scopes = new ConcurrentLinkedDeque<>();
```
Comment (brunobat) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/tracing/OTelTracingProvider.java:48]: How do you know which scopes belong to each request?

Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/tracing/OTelTracingProvider.java:48]: @brunobat The `OTelTracingProvider` is now request-scoped, so every new request creates a new instance of the provider. It means these scopes are attached only to the current request. The root span (parent) is the one created by Quarkus/Vert.x REST instrumentation.

Comment (brunobat) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/tracing/OTelTracingProvider.java:48]: excelent!

**Review by brunobat — APPROVED** [review]

**Comment by ahus1** [comment]
I agree this is an improvement, and that it should be request scoped. I'll merge it in a minute.

Some concerns for a future PR:

* ConcurrentLinkedDeque might not be necessary, as we never concurrently modify this in a request
* OTelTracingProviderFactory creates the SINGLETON on the first tracing request. In tests, there might be multiple factories (although never in a production Keycloak instance). Also the `close()` of the factory will be called on shutdown of Keycloak, which will be a very different CDI context. So I wonder if we can remove the code in the `close()` altogether. I tested it locally, and the shutdown of the OpenTelemetrySDK happens independently anyway.

**Review by ahus1 — APPROVED** [review]

**Comment by mabartos** [comment]
@brunobat @ahus1 Thanks for the review.

> ConcurrentLinkedDeque might not be necessary, as we never concurrently modify this in a request

Yes, probably not, but I wanted to be on the safe side as breaking the scope order might expose some false data about spans. With regard to performance and the sync ops, I would not worry about it so much as the frequency of the traced code (and nested spans) in Keycloak would not be so high in the mid-term future. 

> OTelTracingProviderFactory creates the SINGLETON on the first tracing request. In tests, there might be multiple factories (although never in a production Keycloak instance). Also the close() of the factory will be called on shutdown of Keycloak, which will be a very different CDI context. So I wonder if we can remove the code in the close() altogether. I tested it locally, and the shutdown of the OpenTelemetrySDK happens independently anyway.

Yes, the SINGLETON just stores the `OpenTelemetry` bean (application scoped) for the first tracing request to not "bother" CDI/Arc around it. As stated in Quarkus documentation, beans are not automatically removed when obtained through the programmatic way via `CDI.current()`[1]. I believe it's more the case for request-scoped, or session-scoped beans, so we might assume Quarkus do the proper cleanup - but I explicitly did that.

@ahus1 Not sure if any action is required right now. Or do you want to change it?

[1] https://quarkus.io/guides/cdi-reference#remove_unused_beans

**Comment by mabartos** [comment]
> About the creation: As it is not using volatile and two-step locking for the singleton, it might also be retrieved multiple times or currently form Arc - also that shouldn't lead to problems as those concurrent accesses should be channeled by Arc.

+1

**Pull request merged**